pkg/volume/csi/csi_attacher.go

```diff
@@ -354,6 +341,21 @@ func (c *csiAttacher) MountDevice(spec *volume.Spec, devicePath string, deviceMo
 		fsType = defaultFSType
 	}
 
+	// Save volDataFileName as the very last item before NodeStageVolume, so we
+	// don't need to clean it up in every error path.
+	data := map[string]string{
+		volDataKey.volHandle:  csiSource.VolumeHandle,
+		volDataKey.driverName: csiSource.Driver,
+	}
+	if err := saveVolumeData(dataDir, volDataFileName, data); err != nil {
+		glog.Error(log("failed to save volume info data: %v", err))
+		if err := os.RemoveAll(dataDir); err != nil {
```

Comment:
Do all the error returns above also need to RemoveAll(dataDir)?  Maybe this cleanup should be in a defer function that only runs if err != nil ?

Reply:
`MkdirAll()` is just a few lines above and there is no return in between. I'll move it even closer to `NodeStageVolume` with some comment.